### ISSUE TITLE
Question for supporting kafka-connect version 2.4?

### ISSUE THREAD

**Issue by elakito** [Apicurio/apicurio-registry#557]
Kafka 2.4 API has new method `fromConnectData(String topic, Headers headers, Schema schema, Object value)` in `org.apache.kafka.connect.storage.Converter`. As a result, when trying to use apicurio's current `AvroConverter` with kafka 2.4, it leads to the below exception.
```
java.lang.NoSuchMethodError: org.apache.kafka.connect.storage.Converter.fromConnectData(Ljava/lang/String;Lorg/apache/kafka/common/header/Headers;Lorg/apache/kafka/connect/data/Schema;Ljava/lang/Object;)[B
	at org.apache.kafka.connect.runtime.WorkerSourceTask.lambda$convertTransformedRecord$1(WorkerSourceTask.java:287)
	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndRetry(RetryWithToleranceOperator.java:128)
	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java:162)
	at org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execute(RetryWithToleranceOperator.java:104)
	at org.apache.kafka.connect.runtime.WorkerSourceTask.convertTransformedRecord(WorkerSourceTask.java:287)
	at org.apache.kafka.connect.runtime.WorkerSourceTask.sendRecords(WorkerSourceTask.java:316)
	at org.apache.kafka.connect.runtime.WorkerSourceTask.execute(WorkerSourceTask.java:240)
	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:177)

```
Is there something planned for supporting 2.4 API?
regards, aki

**Comment by EricWittmann** [comment]
@alesj What do you think?  I'd love to hear from the messaging team on this.

**Comment by alesj** [comment]
Impl wise I think this is trivial to support.
It's just a matter which version we (Registry) support.
Or we could hack this with reflection. :-)
(then just need to see how this plays with Quarkus native ...)

**Comment by elakito** [comment]
@alesj Maybe implementing both methods could support both kafka versions? Let me know if you or someone else is looking into it. If no one is looking into it, I'll try to look into it next week. Thanks.

**Comment by EricWittmann** [comment]
If you have some time next week to look into it that would be super helpful, thank you.

**Comment by elakito** [comment]
ok. I'll do.

**Comment by elakito** [comment]
It looks like the original error that I saw came from some conflicting jars that I had in my classpath. For this particular kafka-connnect's Convert interface, there is no compatibility issue with apicurio-regisry binary, as kafka-connect-2.4 Converter's new fromConnnectData has the default implementation. I verified that apicurio-1.2.1.Final indeed works with kafka-connect-2.4.1 in an avro conversion scenario.

The only binary incompatibility in apicurio-registry-1.2.1 against kafka-connect-2.4 seems to be its use of org.apache.kafka.connect.runtime.Worker and org.apache.kafka.connect.runtime.distributed.DistributedHerder in io.apicurio.registry.connector.ConnectorApplication. The 2.4 constructors of these classes have a new argument ConnectorClientConfigOverridePolicy. I think we can use reflection to handle both constructors. I'll look into it to see how this looks.
